token/program-2022/src/extension/transfer_fee/mod.rs

```diff
@@ -32,63 +35,71 @@ pub struct TransferFee {
     pub transfer_fee_basis_points: PodU16,
 }
 impl TransferFee {
+    /// Calculate ceiled division
+    fn ceil_div(dividend: u128, divisor: u128) -> Option<u128> {
+        dividend
+            .checked_add(divisor)?
+            .checked_sub(1)?
+            .checked_div(divisor)
+    }
+
     /// Calculate the transfer fee
     pub fn calculate_fee(&self, pre_fee_amount: u64) -> Option<u64> {
         let transfer_fee_basis_points = u16::from(self.transfer_fee_basis_points) as u128;
         if transfer_fee_basis_points == 0 || pre_fee_amount == 0 {
             Some(0)
         } else {
             let numerator = (pre_fee_amount as u128).checked_mul(transfer_fee_basis_points)?;
-            let mut raw_fee = numerator.checked_div(ONE_IN_BASIS_POINTS)?;
-            let remainder = numerator.checked_rem(ONE_IN_BASIS_POINTS)?;
-            if remainder > 0 {
-                raw_fee = raw_fee.checked_add(1)?;
-            }
-            // guaranteed to be ok
-            let raw_fee = u64::try_from(raw_fee).ok()?;
+            let raw_fee = Self::ceil_div(numerator, ONE_IN_BASIS_POINTS)?
+                .try_into() // guaranteed to be okay
+                .ok()?;
+
             Some(cmp::min(raw_fee, u64::from(self.maximum_fee)))
         }
     }
 
-    /// Calculate the fee that would produce the given output
-    pub fn calculate_inverse_fee(&self, post_fee_amount: u64) -> Option<u64> {
+    /// Calculate the net transfer amount after deducting fees
+    pub fn calculate_post_fee_amount(&self, pre_fee_amount: u64) -> Option<u64> {
+        pre_fee_amount.checked_sub(self.calculate_fee(pre_fee_amount)?)
+    }
+
+    /// Calculate the transfer amount that will result in a specified net transfer amount.
+    ///
+    /// The original transfer amount may not always be unique due to rounding. In this case, the
+    /// smaller amount will be chosen.
+    /// e.g. Both transfer amount 10, 11 with 10% fee rate results in net transfer amount of 9. In
+    /// this case, 10 will be chosen.
+    /// e.g. Fee rate is 100%. In this case, 0 will be chosen.
+    ///
+    /// The original transfer amount may not always exist on large net transfer amounts due to
+    /// overflow. In this case, `None` is returned.
+    /// e.g. The net fee amount is `u64::MAX` with a positive fee rate.
+    pub fn calculate_pre_fee_amount(&self, post_fee_amount: u64) -> Option<u64> {
+        let maximum_fee = u64::from(self.maximum_fee);
         let transfer_fee_basis_points = u16::from(self.transfer_fee_basis_points) as u128;
         if transfer_fee_basis_points == 0 {
+            Some(post_fee_amount)
+        } else if transfer_fee_basis_points == ONE_IN_BASIS_POINTS || post_fee_amount == 0 {
```

Comment:
If post fee amount is 0, that doesn't necessarily mean that the pre fee amount is 0. If I transfer 1 token and the fee is 1bp, that'll have a fee of 1.

Reply:
Yeah, so this is something that I thought about for a bit and is related to one of the commented out tests below as well. The gross amount is not uniquely determined due to rounding. So I thought the function should naturally provide the minimum gross amount that will produce a specified net amount.

Suppose a sender wants a receiver to receive a certain number of tokens after fees. Then it could use this function to calculate how much to send in total before fees are deducted. Then it is natural to always choose the smaller amount since the sender does not want to overpay when possible.

If the post fee amount is 0, then both 0 and 1 should work. If the fee rate is 10,000, then any amount below max fee should also produce 0. In all these cases, the amount 0 is the minimum pre fee amount that will produce 0 post fee amount.

Reply:
We had a discussion about this offline, and you're right, we'll go with this implementation!